### PR TITLE
FOGL-6612: ACL relationship storage changes (Python). Also contains changes for FOGL-6752 and FOGL-6751.

### PR DESCRIPTION
**Pull request by deepanshu-yadav**
Steps

```
export FLEDGE_ROOT=/home/deepanshu/foglamp_workspace/FOGL-6612/fledge

$FLEDGE_ROOT/scripts/fledge kill 

echo YES | $FLEDGE_ROOT/scripts/fledge reset

$FLEDGE_ROOT/scripts/fledge start


curl -sX POST http://localhost:8081/fledge/service -d '{"name":"sine1","type":"south","plugin":"sinusoid","enabled":false,"config":{"assetName":{"value":"asset1"}}}' |jq

curl -sX POST http://localhost:8081/fledge/ACL -d '{"name":"acl1","service":[{"name":"Fledge Core"},{"type":"Core"}],"url":[{"url":"/fledge/category","acl":[{"type":"Core"}]}]}' | jq


curl -sX POST http://localhost:8081/fledge/ACL -d '{"name":"acl2","service":[{"name":"Fledge Core"},{"type":"Core"}],"url":[{"url":"/fledge/category","acl":[{"type":"Core"}]}]}' | jq



curl -sX PUT http://localhost:8081/fledge/schedule/enable -d '{"schedule_name":"sine1"}' |jq
sleep 5
```

Then

```
# This should detach acl1 to sine1
curl -X PUT -H "Content-Type: application/json" -d '{"value": ""}' http://localhost:8081/fledge/category/sine1Security/ACL | jq

# This should attach acl1 to sine1
curl -X PUT -H "Content-Type: application/json" -d '{"value": "acl1"}' http://localhost:8081/fledge/category/sine1Security/ACL | jq

# This should update from acl1 to acl2 for sine1
curl -X PUT -H "Content-Type: application/json" -d '{"value": "acl2"}' http://localhost:8081/fledge/category/sine1Security/ACL | jq

```

Now you can update / delete an ACL then also notification will be sent to service.

```
curl -sX PUT http://localhost:8081/fledge/ACL/acl2 -d '{"service":[{"name":"Fledge Core"},{"type":"Core"}],"url":[{"url":"/fledge/schedule","acl":[{"type":"Core"}]}]}' | jq

curl  -sX DELETE http://localhost:8081/fledge/ACL/acl2 | jq

```

Also we can now attach detach acl to service.

```
curl -X PUT http://localhost:8081/fledge/service/sine1/ACL -d '{"acl_name" : "acl1"}' | jq
curl -X DELETE http://localhost:8081/fledge/service/sine1/ACL | jq

```

### CODE REVIEW

**Comment by deepanshu-yadav** [comment]
Hi @MarkRiddoch  and @pintomax ,

I able to notify the south service but I am getting the following error. 

```
Jul 22 19:06:58 foglamp Fledge[16764] INFO: acl_management: fledge.common.acl_management: Notifying the service sine1 about attachACL
Jul 22 19:06:58 foglamp Fledge sine1[17202]: ERROR: HTTP error while adding child categories: 404: Invalid destination 10
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: Signal 6 (Aborted) trapped:
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (0) 0       0x55c88afbd329 handler(int) + 73---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (1) 1       0x7f434cd7ef10 /lib/x86_64-linux-gnu/libc.so.6(+0x3ef10) [0x7f434cd7ef10]---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (2) 2       0x7f434cd7ee87 gsignal + 199---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (3) 3       0x7f434cd807f1 abort + 321---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (4) 4       0x7f434d3d5957 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x8c957) [0x7f434d3d5957]---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (5) 5       0x7f434d3dbae6 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92ae6) [0x7f434d3dbae6]---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (6) 6       0x7f434d3dbb21 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92b21) [0x7f434d3dbb21]---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (7) 7       0x7f434d3dbd54 /usr/lib/x86_64-linux-gnu/libstdc++.so.6(+0x92d54) [0x7f434d3dbd54]---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (8) 8       0x7f434dbd943e ManagementClient::addChildCategories(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) + 1566---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (9) 9       0x7f434d8f2fe9 ServiceAuthHandler::createSecurityCategories(ManagementClient*) + 981---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (10) 10      0x55c88afc173c SouthService::start(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&, unsigned short) + 3228---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (11) 11      0x55c88afaed1d main + 605---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (12) 12      0x7f434cd61c87 __libc_start_main + 231---------
Jul 22 19:06:58 foglamp Fledge sine1[17202]: FATAL: (13) 13      0x55c88afaf84a _start + 42---------

```

So I think you can continue from here on. 
Also In order to create a config item of type ACL use the method mentioned in above comment.

File: python/fledge/common/microservice_management_client/microservice_management_client.py [hunk 1]
```diff
@@ -298,6 +300,28 @@ def update_configuration_item(self, category_name, config_item, category_data):
         response = json.loads(res)
         return response
 
+    def update_service_for_acl_change_security(self, acl, reason):
+        assert reason in ["attachACL", "detachACL", "updateACL"]
+        url = "/fledge/security"
+        payload = {
+            "reason": reason,
+            "argument": acl
```
Comment (deepanshu-yadav) [inline python/fledge/common/microservice_management_client/microservice_management_client.py:308]: Hi please confirm that you are expecting this payload in your implementation.

Comment (pintomax) [reply python/fledge/common/microservice_management_client/microservice_management_client.py:308]: @deepanshu-yadav In the PR #724 for FOGL-611, I have tested:

curl -v -X PUT 127.0.0.1:41615/fledge/security -d '{"reason" : "attachACL", "argument" : "ACL1"}'

**Comment by deepanshu-yadav** [comment]
> @deepanshu-yadav by commenting out all code in
> 
> _notify_service_about_acl_change
> 
> about MicroserviceManagementClient, then there is no crash

Fixed the previous crash. Now the I am facing the following issue. 
I am able to send a PUT request to fledge/security but not getting anything in return. So waiting indefinitely.

```
Jul 25 12:42:50 foglamp Fledge[6453] DEBUG: configuration_manager: fledge.common.configuration_manager: check if acl item there is True for parent sine1
Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: The tuple (sine1, service, acl1) already exists in acl usage table.
Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: Notifying the service sine1 about attachACL
Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: The host and port is localhost 33753 and entity is sine1
Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: Connect established with sine1 at localhost and port 33753
Jul 25 12:42:50 foglamp Fledge[6453] INFO: microservice_management_client: fledge.common.microservice_management_client.microservice_management_client: Requesting end point /fledge/security with payload {'reason': 'attachACL', 'argument': 'acl1'}

```

**Comment by MarkRiddoch** [comment]
> > @deepanshu-yadav by commenting out all code in
> > _notify_service_about_acl_change
> > about MicroserviceManagementClient, then there is no crash
> 
> Fixed the previous crash. Now the I am facing the following issue. I am able to send a PUT request to fledge/security but not getting anything in return. So waiting indefinitely.
> 
> ```
> Jul 25 12:42:50 foglamp Fledge[6453] DEBUG: configuration_manager: fledge.common.configuration_manager: check if acl item there is True for parent sine1
> Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: The tuple (sine1, service, acl1) already exists in acl usage table.
> Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: Notifying the service sine1 about attachACL
> Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: The host and port is localhost 33753 and entity is sine1
> Jul 25 12:42:50 foglamp Fledge[6453] INFO: acl_management: fledge.common.acl_management: Connect established with sine1 at localhost and port 33753
> Jul 25 12:42:50 foglamp Fledge[6453] INFO: microservice_management_client: fledge.common.microservice_management_client.microservice_management_client: Requesting end point /fledge/security with payload {'reason': 'attachACL', 'argument': 'acl1'}
> ```

Something is very wrong here. For one thing you should never wait indefinitely, but also if you are sending to the right port and address then you should get a response, even if it is only an error response to say you have the wrong path URL. Is the south service there and listening on that port? Is the south service crashing?

**Comment by pintomax** [comment]
Latest commit of FOGL-6612_API branch is needed for /fledge/security operation

Get new updates from that branch and run make in order to compile needed C++ code

**Comment by pintomax** [comment]
@deepanshu-yadav 

ACL1 created with Fledge GUI

SINSecurity has item ACL with value = ""


curl -X PUT 127.0.0.1:8081/fledge/category/SIN1Security/ACL -d '{"value" : "ACL1"}'

This command calls service SIN1 on its management port PUT /fledge/security

The C++ code then calls CORE management port for /fledge/ACL/ACL1 and that is being blocked

At this point any call to Fledge port 8081 or management port are blocked:

curl 127.0.0.1:8081/fledge/category

etc

it seems a Python lock issue in the HTTP server

**Comment by deepanshu-yadav** [comment]
> @deepanshu-yadav
> 
> ACL1 created with Fledge GUI
> 
> SINSecurity has item ACL with value = ""
> 
> curl -X PUT 127.0.0.1:8081/fledge/category/SIN1Security/ACL -d '{"value" : "ACL1"}'
> 
> This command calls service SIN1 on its management port PUT /fledge/security
> 
> The C++ code then calls CORE management port for /fledge/ACL/ACL1 and that is being blocked
> 
> At this point any call to Fledge port 8081 or management port are blocked:
> 
> curl 127.0.0.1:8081/fledge/category
> 
> curl 127.0.0.1:42981/fledge/service/ping
> 
> etc
> 
> it seems a Python lock issue in the HTTP server

Resolved it was a deadblock between python and c++ side both trying to send  some http request through management client.

**Review by pintomax — APPROVED** [review]

**Review by pintomax — DISMISSED** [review]

**Comment by pintomax** [comment]
Updating an ACL, ACL1, already associated to service SIN1, via PUT /fledge/ACL/ACL1, now calls
service security endpoint /PUT /fledge/security with payload {"reloadACL" : "ACL1"}

**Review by MarkRiddoch — DISMISSED** [review]

File: C/services/common/service_security.cpp [hunk 2]
```diff
@@ -772,3 +720,61 @@ static void bearer_token_refresh_thread(void *data)
...
+
+	string r = reason.getReason();
+
+	// Lock config
+	lock_guard<mutex> cfgLock(m_mtx_config);
+
+	if (r == "attachACL")
+	{
+		// Fetch and load ACL
+		m_service_acl = m_mgtClient->getACL(reason.getArgument());
```
Comment (deepanshu-yadav) [inline C/services/common/service_security.cpp:751]: Hi @pintomax , 
There are some TODOS here.

**Review by pintomax — APPROVED** [review]

**Pull request merged**